packages/source-map/lib/sourceMap.ts

```diff
@@ -11,76 +11,151 @@ export interface Mapping<Data = unknown> {
+	private sourceCodeOffsetsStorage: MapStorage<Data> | undefined;
+	private generatedCodeOffsetsStorage: MapStorage<Data> | undefined;
 
-	constructor(public readonly mappings: Mapping<Data>[]) { }
+	constructor(public readonly mappings: Mapping<Data>[]) {
+	}
 
-	getSourceStartEnd(generatedStart: number, generatedEnd: number) {
-		return this.findMatchingStartEnd(generatedStart, generatedEnd, 'generatedOffsets', 'sourceOffsets');
+	getSourceStartEnd(generatedStart: number, generatedEnd: number, filter?: (data: Data) => boolean): Generator<[start: number, end: number, Mapping<Data>]> {
+		return this.findMatchingStartEnd(generatedStart, generatedEnd, 'generatedOffsets', filter);
 	}
 
-	getGeneratedStartEnd(sourceStart: number, sourceEnd: number) {
-		return this.findMatchingStartEnd(sourceStart, sourceEnd, 'sourceOffsets', 'generatedOffsets');
+	getGeneratedStartEnd(sourceStart: number, sourceEnd: number, filter?: (data: Data) => boolean): Generator<[start: number, end: number, Mapping<Data>]> {
+		return this.findMatchingStartEnd(sourceStart, sourceEnd, 'sourceOffsets', filter);
 	}
 
-	* findMatchingStartEnd(start: number, end: number, fromRange: CodeRangeKey, toRange: CodeRangeKey) {
-		for (const [mappedStart, mapping] of this.findMatching(start, fromRange, toRange)) {
-			const mappedEnd = translateOffset(end, mapping[fromRange], mapping[toRange], getLengths(mapping, fromRange), getLengths(mapping, toRange));
-			if (mappedEnd !== undefined) {
-				yield [mappedStart, mappedEnd, mapping] as const;
-			}
-		};
+	getSourceOffsets(generatedOffset: number, filter?: (data: Data) => boolean): Generator<[offset: number, Mapping<Data>]> {
+		return this.findMatchingOffsets(generatedOffset, 'generatedOffsets', filter);
 	}
 
-	getSourceOffsets(generatedOffset: number) {
-		return this.findMatching(generatedOffset, 'generatedOffsets', 'sourceOffsets');
+	getGeneratedOffsets(sourceOffset: number, filter?: (data: Data) => boolean): Generator<[offset: number, Mapping<Data>]> {
+		return this.findMatchingOffsets(sourceOffset, 'sourceOffsets', filter);
 	}
 
-	getGeneratedOffsets(sourceOffset: number) {
-		return this.findMatching(sourceOffset, 'sourceOffsets', 'generatedOffsets');
+	* findMatchingOffsets(offset: number, fromRange: CodeRangeKey, filter?: (data: Data) => boolean): Generator<[number, Mapping<Data>]> {
+		const toRange: CodeRangeKey = fromRange == 'sourceOffsets' ? 'generatedOffsets' : 'sourceOffsets'
+		for (const match of this.getStorageBasedOnRange(fromRange).findMatchingMappingIndexes(offset, filter)) {
+			yield [mapOffset(match, offset, fromRange, toRange), match[0]];
+		}
 	}
 
-	* findMatching(offset: number, fromRange: CodeRangeKey, toRange: CodeRangeKey) {
-		const memo = this.getMemoBasedOnRange(fromRange);
-		if (memo.offsets.length === 0) {
-			return;
+	* findMatchingStartEnd(start: number, end: number, fromRange: CodeRangeKey, filter?: (data: Data) => boolean): Generator<[start: number, end: number, Mapping<Data>]> {
```

Comment:
Moving range matching logic to the source map ensures that we have consistency across the codebase